packages/babel-plugin-transform-typescript/src/index.ts

```diff
@@ -56,6 +57,7 @@ export default declare((api, opts) => {
     jsxPragma = "React.createElement",
     jsxPragmaFrag = "React.Fragment",
     onlyRemoveTypeImports = false,
+    optimizeConstEnums = false,
```

Comment:
Can we default it to `true` in next minor? I know we have `onlyRemoveTypeImports` default to `true` in Babel 8, but supporting new language features should be opt-in by default, like we did for other ES features.

Reply:
I'm not sure we should.

In TypeScript with `--isolatedModules`, code like this works and logs `x`:
```ts
const enum A {
  x
}

// @ts-expect-error
console.log(A[A.x]);

export {};
```

However, with `optimizeConstEnums` it stops working because we don't ignore `const` anymore.

For the default behavior, I'd prefer to match `--isolatedModules` as closely as possible where it's easy to do so.

Reply:
Interesting, I think it should be a TS issue, they could have optimized the const enums here.

Reply:
[TypeScript docs](https://www.typescriptlang.org/tsconfig#references-to-const-enum-members) mentions why ambient `const enum`s (ie. `declare const enum`) don't work with `isolatedModules`.

When `isolatedModules: false`, it seems `const enum` and `declare const enum` function the same.

When `isolatedModules: true`, it seems `const enum` transpiles as if it was just `enum` since you get the reverse-mapped object. `declare const enum` transpiles into the usual empty output, because it's expected to be inlined. However, the `isolatedModules: true` will produce error:

```
Cannot access ambient const enums when the '--isolatedModules' flag is provided.(2748)
```

No idea why they only document `declare const enum` and not `const enum`. But I assume their reasoning of converting `const enum` to just `enum` is to have it work across modules. And I guess they treat `declare const enum` as an error since you are explicitly saying it is ambient.

Reply:
The default Babel behavior is to match `--isolatedModules` ([as mentioned in our docs](https://babeljs.io/docs/en/babel-plugin-transform-typescript#caveats) at the 4th caveat).

Thus, this PR makes it so that by default `const enum`s are compiled to "normal" enums, which also matches the default behavior of your `babel-plugin-const-enum` plugin. It's implemented in [`933ba59` (#13324)](https://github.com/babel/babel/pull/13324/commits/933ba59f4e1c12b8ec93b52fb980328d92d4dbcb).

With `optimizeConstEnums: true` the plugin seems to correctly work: https://github.com/babel/babel/pull/13324/files#diff-d0d0efefe2a94935f80a8f399bbca77e2a0fbbc8e7a3ba2a1a2f07eff7bcb0de. When `optimizeConstEnums: false`, it produces some broken output:
```js
declare const enum Animal {
  Dog,
  Cat
}
let x: Animal = Animal.Dog;

// becomes

let x = Animal.Dog; 
```
however, I'm not too much concerned about it because:
- It's exactly the same output that TS generates with `--isolatedMoudles`
- TypeScript will report a type-checking error saying that `declare const enum` are not supported.